### PR TITLE
[PluggableDevice] TensorList support with DEVICE_DEFAULT

### PR DESCRIPTION
**Pull request by i8run**
Use stream apis to enable the TensorList with DEVICE_DEFAULT including memzero, memset and memset32 and it's only enabled with pluggable device.

### CODE REVIEW

**Comment by i8run** [comment]
@penpornk @saxenasaurabh Could you please help to review this PR? It implements the TensorList with stream APIs instead of exposing C APIs when using pluggable device. Thanks.

**Review by penpornk — CHANGES_REQUESTED** [review]
I'm so sorry for the delay! I have a few quick comments. Will take a closer look later.

File: tensorflow/core/kernels/list_kernels.cc [hunk 1]
```diff
@@ -159,16 +159,12 @@ class EmptyTensorList : public OpKernel {
 REGISTER_KERNEL_BUILDER(Name("EmptyTensorList").Device(DEVICE_CPU),
                         EmptyTensorList);
 
-#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
-
 REGISTER_KERNEL_BUILDER(Name("EmptyTensorList")
-                            .Device(DEVICE_GPU)
```
[the remark this answers is not in the document]
Comment (i8run) [reply tensorflow/core/kernels/list_kernels.cc]: Done. Add `DEVICE_DEFAULT` below the GPU registration.

File: tensorflow/core/kernels/list_kernels.h [hunk 2]
```diff
@@ -531,8 +609,7 @@ class TensorListSplit : public OpKernel {
       // prevent this.
       Tensor aligned;
       OP_REQUIRES_OK(c, c->allocate_temp(tmp.dtype(), tmp.shape(), &aligned));
-      aligned.flat<T>().device(c->eigen_device<Device>()) =
-          tmp.unaligned_flat<T>();
+      CopyTensor<Device, T>(c, tmp, aligned);
```
[the remark this answers is not in the document]
Comment (i8run) [reply tensorflow/core/kernels/list_kernels.h]: I use a function pointer to implement this in kernels. Do you think it's OK ?

File: tensorflow/core/kernels/list_kernels.h [hunk 3]
```diff
@@ -55,6 +56,75 @@ Status ForwardInputOrCreateNewList(OpKernelContext* c, int32_t input_index,
...
                                    TensorList** output_list);
 
+template <typename Device, typename T>
+void SetZero(OpKernelContext* ctx, Tensor& tensor) {
+  auto device_context = ctx->op_device_context();
+  if (device_context && device_context->IsPluggableDevice()) {
+    auto ptr =
+        se::DeviceMemoryBase(tensor.flat<T>().data(), tensor.TotalBytes());
+    auto stream = ctx->op_device_context()->stream();
+    CHECK_EQ(true, stream->ThenMemZero(&ptr, tensor.TotalBytes()).ok());
```
[the remark this answers is not in the document]
Comment (i8run) [reply tensorflow/core/kernels/list_kernels.h]: Done.

File: tensorflow/core/kernels/list_kernels.h [hunk 4]
```diff
@@ -706,8 +810,15 @@ Status Scatter(OpKernelContext* c, const Tensor& value, const Tensor& indices,
...
-    aligned.flat<T>().device(c->eigen_device<Device>()) =
-        tmp.unaligned_flat<T>();
+    void (*copy_tensor)(OpKernelContext * ctx, Tensor&, Tensor&);
+    bool is_pluggable_device =
+        c->op_device_context() && c->op_device_context()->IsPluggableDevice();
+    if (is_pluggable_device) {
+      copy_tensor = &CopyTensorPluggableDevice<T>;
+    } else {
+      copy_tensor = &CopyTensor<Device, T>;
+    }
```
Comment (penpornk) [inline tensorflow/core/kernels/list_kernels.h:820]: This `if` should be outside the for loop.

**Review by penpornk — APPROVED** [review]
Thank you for the changes. I have some more comments but I think I'll just pull the PR and make additional changes internally given that the TF 2.7 branch cut is tomorrow. (Sorry for the delay!)

**Pull request merged**